File: nn/nn.ts
import { Tensor } from './tensor';

export abstract class Module {
  parameters(): Tensor[] {
    const params: Tensor[] = [];
    for (const v of Object.values(this)) {
      if (v instanceof Tensor && v.requires_grad) params.push(v);
      else if (v instanceof Module) params.push(...v.parameters());
      else if (Array.isArray(v) && v.some(item => item instanceof Module)) {
        for (const item of v) if (item instanceof Module) params.push(...item.parameters());
      }
    }
    return params;
  }
}

export class Linear extends Module {
  weight: Tensor;
  bias: Tensor | null;

  constructor(in_features: number, out_features: number, bias = false) {
    super();
    const std = 0.08;
    this.weight = Tensor.from(
      Array(out_features).fill(0).map(() =>
        Array(in_features).fill(0).map(() => (Math.random() * 2 - 1) * std)
      ),
      true
    );
    this.bias = bias
      ? Tensor.from([Array(out_features).fill(0)], true)
      : null;
  }

  forward(x: Tensor): Tensor {
    const out = x.matmul(this.weight.transpose());
    if (this.bias) return out.add(this.bias);
    return out;
  }

  __call__(x: Tensor): Tensor {
    return this.forward(x);
  }
}

export class Embedding extends Module {
  weight: Tensor;

  constructor(num_embeddings: number, embedding_dim: number) {
    super();
    this.weight = Tensor.from(
      Array(num_embeddings).fill(0).map(() =>
        Array(embedding_dim).fill(0).map(() => (Math.random() * 2 - 1) * 0.08)
      ),
      true
    );
  }

  forward(indices: Tensor): Tensor {
    const idx = indices.data.map(row => row.map(v => Math.round(Number(v))));
    const out: number[][] = idx.flatMap(row =>
      row.map(i => this.weight.data[i] ?? this.weight.data[0] ?? [])
    );
    const result = new Tensor(out, [this.weight]);
    (result as any)._backward = () => {
      const grad = result.grad;
      for (const row of idx) {
        for (const i of row) {
          for (let j = 0; j < (this.weight.grad[i]?.length ?? 0); j++) {
            this.weight.grad[i][j] += grad[idx.indexOf(row) as number]?.[j] ?? 0;
          }
        }
      }
    };
    return result;
  }

  __call__(indices: Tensor): Tensor {
    return this.forward(indices);
  }
}

export class RMSNorm extends Module {
  eps = 1e-5;
  scale: Tensor;

  constructor(dim: number) {
    super();
    this.scale = Tensor.from([Array(dim).fill(1)], false);
  }

  forward(x: Tensor): Tensor {
    const ms = x.data.map(row => {
      const m2 = row.reduce((s, v) => s + v * v, 0) / row.length;
      return Math.sqrt(m2 + this.eps);
    });
    const out = x.data.map((row, i) => row.map(v => v / ms[i]));
    const result = new Tensor(out, [x]);
    result.requires_grad = x.requires_grad;
    result._backward = () => {
      if (!x.requires_grad) return;
      const inv_std = ms.map(m => 1 / m);
      for (let i = 0; i < x.grad.length; i++) {
        for (let j = 0; j < x.grad[i].length; j++) {
          x.grad[i][j] += result.grad[i]?.[j] ?? 0 * inv_std[i];
        }
      }
    };
    return result;
  }

  __call__(x: Tensor): Tensor {
    return this.forward(x);
  }
}

export class Adam {
  params: Tensor[];
  lr: number;
  beta1: number;
  beta2: number;
  eps: number;
  m: number[][][] = [];
  v: number[][][] = [];
  t = 0;

  constructor(params: Tensor[], lr = 0.01, betas: [number, number] = [0.85, 0.99], eps = 1e-8) {
    this.params = params;
    this.lr = lr;
    this.beta1 = betas[0];
    this.beta2 = betas[1];
    this.eps = eps;
    this.m = params.map(p => p.data.map(row => row.map(() => 0)));
    this.v = params.map(p => p.data.map(row => row.map(() => 0)));
  }

  step(): void {
    this.t++;
    for (let i = 0; i < this.params.length; i++) {
      const p = this.params[i];
      for (let j = 0; j < p.data.length; j++) {
        for (let k = 0; k < p.data[j].length; k++) {
          const g = p.grad[j]?.[k] ?? 0;
          this.m[i][j][k] = this.beta1 * this.m[i][j][k] + (1 - this.beta1) * g;
          this.v[i][j][k] = this.beta2 * this.v[i][j][k] + (1 - this.beta2) * g * g;
          const m_hat = this.m[i][j][k] / (1 - this.beta1 ** this.t);
          const v_hat = this.v[i][j][k] / (1 - this.beta2 ** this.t);
          p.data[j][k] -= this.lr * m_hat / (Math.sqrt(v_hat) + this.eps);
        }
      }
    }
  }

  zeroGrad(): void {
    for (const p of this.params) p.zeroGrad();
  }
}

File: nn/nn.py
"""
Neural network layers and optimizers.
Module: base class for all neural network components
Linear: fully connected layer (y = x @ W + b)
Embedding: token embedding lookup table
RMSNorm: Root Mean Square Normalization
Adam: Adaptive Moment Estimation optimizer
"""

from __future__ import annotations

import numpy as np
from .tensor import Tensor


def mse_loss(input: Tensor, target: Tensor) -> Tensor:
    """Mean Squared Error loss.

    FIX: original _backward read diff.grad, which is always zero because
    diff = input - target is an intermediate node created outside the
    autograd graph registered to `out`. backward() never traverses diff.
    Now computes the gradient directly from diff.data (the stored values).
    """
    diff = input - target
    out = Tensor(np.mean(diff.data ** 2), (input, target), input.requires_grad or target.requires_grad)

    def _backward() -> None:
        N = np.prod(diff.data.shape)
        input.grad  += out.grad * 2 * diff.data / N
        target.grad += out.grad * -2 * diff.data / N

    out._backward = _backward
    return out


class Module:
    """
    Base class for all neural network modules.
    Provides parameter collection via recursive traversal.
    """

    def parameters(self) -> list[Tensor]:
        """
        Recursively collects all Tensors with requires_grad=True.
        Traverses Module attributes, lists of Modules, and direct Tensors.
        """
        params: list[Tensor] = []
        for v in self.__dict__.values():
            if isinstance(v, Tensor) and v.requires_grad:
                params.append(v)
            elif isinstance(v, Module):
                params.extend(v.parameters())
            elif isinstance(v, (list, tuple)):
                for item in v:
                    if isinstance(item, Module):
                        params.extend(item.parameters())
        return params


class Sequential(Module):
    """Sequential container - applies modules in order."""

    def __init__(self, *modules: Module) -> None:
        self.modules: tuple[Module, ...] = modules

    def __call__(self, x: Tensor) -> Tensor:
        for m in self.modules:
            x = m(x)
        return x


class ReLU(Module):
    """ReLU activation layer."""

    def __call__(self, x: Tensor) -> Tensor:
        return x.relu()


class Tanh(Module):
    """Tanh activation layer."""

    def __call__(self, x: Tensor) -> Tensor:
        return x.tanh()


class Linear(Module):
    """
    Fully connected linear transformation: y = x @ W + b
    Weight shape: (in_features, out_features)
    Bias shape: (out_features,) if bias=True, else None
    """

    def __init__(self, in_features: int, out_features: int, bias: bool = False) -> None:
        std = 0.08
        self.weight = Tensor(np.random.normal(0, std, (in_features, out_features)), requires_grad=True)
        self.bias: Tensor | None = Tensor(np.zeros((out_features,)), requires_grad=True) if bias else None

    def __call__(self, x: Tensor) -> Tensor:
        out = x @ self.weight
        if self.bias is not None:
            out = out + self.bias
        return out


class Embedding(Module):
    """
    Token embedding lookup table.
    Maps integer indices to embedding vectors.
    Gradient accumulation uses np.add.at to handle duplicate indices correctly.
    """

    def __init__(self, num_embeddings: int, embedding_dim: int) -> None:
        self.weight = Tensor(np.random.normal(0, 0.08, (num_embeddings, embedding_dim)), requires_grad=True)

    def __call__(self, indices: Tensor) -> Tensor:
        idx = indices.data.astype(int) if isinstance(indices, Tensor) else indices
        out_data = self.weight.data[idx]
        out = Tensor(out_data, (self.weight,), requires_grad=True)

        def _backward() -> None:
            np.add.at(self.weight.grad, idx, out.grad)

        out._backward = _backward
        return out


class RMSNorm(Module):
    """
    Root Mean Square Normalization: x / RMS(x)
    RMS(x) = sqrt(mean(x^2) + eps)
    Cheaper than LayerNorm (no mean subtraction), used in LLaMA and modern LLMs.
    """

    def __init__(self, dim: int, eps: float = 1e-5) -> None:
        self.eps: float = eps
        self.scale = Tensor(np.ones(dim), requires_grad=False)

    def __call__(self, x: Tensor) -> Tensor:
        ms = np.mean(x.data ** 2, axis=-1, keepdims=True) + self.eps
        inv_std = ms ** -0.5
        out_data = x.data * inv_std
        out = Tensor(out_data, (x,), requires_grad=True)

        def _backward() -> None:
            N = x.data.shape[-1]
            dx = (out.grad * inv_std) - (
                x.data * inv_std ** 3 * np.sum(out.grad * x.data, axis=-1, keepdims=True) / N
            )
            x.grad += dx

        out._backward = _backward
        return out


class Adam(Module):
    """
    Adam optimizer (Adaptive Moment Estimation).

    Maintains two exponential moving averages per parameter:
    - m: first moment (gradient momentum), biased toward zero
    - v: second moment (squared gradient momentum), biased toward zero

    Update rule with bias correction:
        m_hat = m / (1 - beta1^t)
        v_hat = v / (1 - beta2^t)
        theta = theta - lr * m_hat / (sqrt(v_hat) + eps)
    """

    def __init__(
        self,
        params: list[Tensor],
        lr: float = 0.01,
        betas: tuple[float, float] = (0.85, 0.99),
        eps: float = 1e-8,
    ) -> None:
        self.params: list[Tensor] = params
        self.lr: float = lr
        self.beta1: float
        self.beta2: float
        self.beta1, self.beta2 = betas
        self.eps: float = eps
        self.m: list[np.ndarray] = [np.zeros_like(p.data) for p in params]
        self.v: list[np.ndarray] = [np.zeros_like(p.data) for p in params]
        self.t: int = 0

    def step(self) -> None:
        """Performs one optimization step."""
        self.t += 1
        for i, p in enumerate(self.params):
            self.m[i] = self.beta1 * self.m[i] + (1 - self.beta1) * p.grad
            self.v[i] = self.beta2 * self.v[i] + (1 - self.beta2) * (p.grad ** 2)
            m_hat = self.m[i] / (1 - self.beta1 ** self.t)
            v_hat = self.v[i] / (1 - self.beta2 ** self.t)
            p.data -= self.lr * m_hat / (np.sqrt(v_hat) + self.eps)

    def zero_grad(self) -> None:
        """Resets gradients of all parameters."""
        for p in self.params:
            p.zero_grad()